projects/cdk/schematics/ng-update/v4/steps/migrate-proprietary.ts

```diff
@@ -0,0 +1,150 @@
+/// <reference lib="es2021" />
+import type {Rule, SchematicContext, Tree} from '@angular-devkit/schematics';
+import {NodePackageInstallTask} from '@angular-devkit/schematics/tasks';
+import {updateWorkspace} from '@schematics/angular/utility/workspace';
+import {
+    addPackageJsonDependency,
+    getActiveProject,
+    getPackageJsonDependency,
+    getSourceFiles,
+    removePackageJsonDependency,
+    saveActiveProject,
+} from 'ng-morph';
+
+import {TUI_VERSION} from '../../../../constants/version';
+import {tuiIsString} from '../../../../utils/miscellaneous/is-string';
+import type {TuiSchema} from '../../../ng-add/schema';
+import {infoLog, REPLACE_SYMBOL, SMALL_TAB_SYMBOL} from '../../../utils/colored-log';
+import {getProjectTargetOptions} from '../../../utils/get-project-target-options';
+import {getProjects} from '../../../utils/get-projects';
+import type {Asset} from '../../interfaces';
+import {getFileSystem} from '../../utils/get-file-system';
+
+export function migrateProprietary(options: TuiSchema) {
+    return (tree: Tree, _: SchematicContext): Rule => {
+        try {
+            if (getPackageJsonDependency(tree, '@taiga-ui/proprietary-core')?.version) {
+                !options['skip-logs'] &&
+                    infoLog(
+                        `${SMALL_TAB_SYMBOL}${REPLACE_SYMBOL} migrating proprietary...`,
+                    );
+
+                replaceScopePackages(tree);
+
+                return replaceAngularJsonStyles(options, tree);
+            }
+        } catch {
+            // noop
+        }
+
+        return () => {};
+    };
+}
+
+function replaceScopePackages(tree: Tree): void {
+    const fileSystem = getFileSystem(tree);
```

Comment:
getFileSystem (setActiveProject / saveActiveProject) is an expensive operation.
Why do we need call it (twice)? can we use filesystem from `main`?

Reply:
fixed